File: src/108.convert_sorted_array_to_binary_search_tree.kt
fun main() {
    println(sortedArrayToBST(intArrayOf(-10, -3, 0, 5, 9)))
    println(sortedArrayToBST(intArrayOf(1, 3)))
}

fun sortedArrayToBST(nums: IntArray): TreeNode? {
    if (nums.isEmpty()) {
        return null
    }

    val mid = nums.size / 2
    val root = TreeNode(nums[mid])

    root.left = sortedArrayToBST(nums.take(mid).toIntArray())
    root.right = sortedArrayToBST(nums.drop(mid+1).toIntArray())

    return root
}

File: lib/108.convert_sorted_array_to_binary_search_tree.dart
import 'other/tree_node.dart';

void main() {
  print(sortedArrayToBST([-10, -3, 0, 5, 9]));
  print(sortedArrayToBST([1, 3]));
}

/// todo
TreeNode? sortedArrayToBST(List<int> nums) {
  if (nums.isEmpty) {
    return null;
  }

  final mid = nums.length ~/ 2;
  final root = TreeNode(nums[mid]);

  root.left = sortedArrayToBST(nums.sublist(0 , mid));
  root.right = sortedArrayToBST(nums.sublist(mid + 1));

  return root;
}
